File: apps/server/src/server/services/run-log.ts
// The in-memory record of recent runs, so /api/runs/:id and the exports can serve rows the
// on-disk history deliberately drops. Bounded: the oldest run falls out once the cap is reached.

import type { RunRecord } from "@perch/protocol";

export const DEFAULT_MAX_RUNS = 50;

/** A copy of the record whose statement results carry counts but no row data. */
export function stripRows(record: RunRecord): RunRecord {
  if (!record.results) return { ...record };
  return { ...record, results: record.results.map((result) => ({ ...result, rows: [] })) };
}

export class RunLog {
  readonly maxRuns: number;
  private readonly runs = new Map<string, RunRecord>();

  constructor(maxRuns: number = DEFAULT_MAX_RUNS) {
    this.maxRuns = maxRuns;
  }

  /**
   * Remembers a run and evicts the oldest ones. Runs are remembered in the order they start, and
   * `Map` preserves insertion order, so the first key is always the oldest run — no scan needed.
   */
  remember(record: RunRecord): void {
    this.runs.set(record.id, record);
    while (this.runs.size > this.maxRuns) {
      const oldest = this.runs.keys().next().value;
      if (oldest === undefined || oldest === record.id) break;
      this.runs.delete(oldest);
    }
  }

  get(runId: string): RunRecord | undefined {
    return this.runs.get(runId);
  }

  /** Newest first, without result rows (rows come from `get`). */
  list(): RunRecord[] {
    return [...this.runs.values()]
      .sort((a, b) => b.startedAt.localeCompare(a.startedAt))
      .map(stripRows);
  }
}

File: apps/server/server/runlog.go
package server

import (
	"sort"
	"sync"

	"perch/protocol"
)

const maxRuns = 50

// RunLog is the in-memory record of recent runs, so /api/runs/:id and the exports can serve rows
// the on-disk history deliberately drops. Bounded: the oldest run falls out at the cap.
type RunLog struct {
	mu    sync.Mutex
	runs  map[string]*protocol.RunRecord
	order []string
}

func NewRunLog() *RunLog {
	return &RunLog{runs: map[string]*protocol.RunRecord{}}
}

func (l *RunLog) Remember(record *protocol.RunRecord) {
	l.mu.Lock()
	defer l.mu.Unlock()
	if _, exists := l.runs[record.ID]; !exists {
		l.order = append(l.order, record.ID)
	}
	l.runs[record.ID] = record
	for len(l.order) > maxRuns {
		oldest := l.order[0]
		if oldest == record.ID {
			break
		}
		l.order = l.order[1:]
		delete(l.runs, oldest)
	}
}

func (l *RunLog) Get(runID string) (protocol.RunRecord, bool) {
	l.mu.Lock()
	defer l.mu.Unlock()
	record, ok := l.runs[runID]
	if !ok {
		return protocol.RunRecord{}, false
	}
	return *record, true
}

// List is newest first, without result rows — those come from Get.
func (l *RunLog) List() []protocol.RunRecord {
	l.mu.Lock()
	out := make([]protocol.RunRecord, 0, len(l.runs))
	for _, record := range l.runs {
		out = append(out, stripRows(*record))
	}
	l.mu.Unlock()

	sort.Slice(out, func(i, j int) bool { return out[i].StartedAt > out[j].StartedAt })
	return out
}

// stripRows copies the record with statement results that carry counts but no row data.
func stripRows(record protocol.RunRecord) protocol.RunRecord {
	if record.Results == nil {
		return record
	}
	slim := make([]protocol.StatementResult, len(record.Results))
	for i, result := range record.Results {
		result.Rows = make([]protocol.Row, 0)
		slim[i] = result
	}
	record.Results = slim
	return record
}
